File: app/src/main/java/com/github/willjgriff/ethereumwallet/ui/send/SendController.java
package com.github.willjgriff.ethereumwallet.ui.send;

import android.support.annotation.NonNull;
import android.support.design.widget.TextInputLayout;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;

import com.github.willjgriff.ethereumwallet.R;
import com.github.willjgriff.ethereumwallet.mvp.BaseMvpController;
import com.github.willjgriff.ethereumwallet.ui.navigation.NavigationToolbarListener;
import com.github.willjgriff.ethereumwallet.ui.send.di.SendInjector;
import com.github.willjgriff.ethereumwallet.ui.send.mvp.SendPresenter;
import com.github.willjgriff.ethereumwallet.ui.send.mvp.SendView;
import com.jakewharton.rxbinding2.view.RxView;
import com.jakewharton.rxbinding2.widget.RxTextView;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.Observable;

/**
 * Created by Will on 29/01/2017.
 */

public class SendController extends BaseMvpController<SendView, SendPresenter> implements SendView {

	@BindView(R.id.controller_send_recipient_address)
	TextInputLayout mRecipientAddress;
	@BindView(R.id.controller_send_amount)
	TextInputLayout mSendAmount;
	@BindView(R.id.controller_send_password)
	TextInputLayout mAccountPassword;
	@BindView(R.id.controller_send_send_ether)
	Button mSendEtherButton;

	@Inject
	SendPresenter mSendPresenter;

	@Override
	protected SendView getMvpView() {
		return this;
	}

	@Override
	protected SendPresenter createPresenter() {
		return mSendPresenter;
	}

	public SendController() {
		SendInjector.INSTANCE.getComponent().inject(this);
	}

	@NonNull
	@Override
	protected View onCreateView(@NonNull LayoutInflater inflater, @NonNull ViewGroup container) {
		View view = inflater.inflate(R.layout.controller_send, container, false);
		ButterKnife.bind(this, view);

		setupToolbarTitle();
		setInputObservables();

		return view;
	}

	private void setupToolbarTitle() {
		if (getTargetController() instanceof NavigationToolbarListener) {
			((NavigationToolbarListener) getTargetController())
				.setToolbarTitle(getApplicationContext().getString(R.string.controller_send_title));
		}
	}

	private void setInputObservables() {
		Observable<CharSequence> recipientAddressObservable = RxTextView.textChanges(mRecipientAddress.getEditText());
		Observable<CharSequence> sendAmountObservable = RxTextView.textChanges(mSendAmount.getEditText());
		Observable<CharSequence> accountPasswordObservable = RxTextView.textChanges(mAccountPassword.getEditText());
		Observable<Object> sendObservable = RxView.clicks(mSendEtherButton);
		getPresenter().setObservables(recipientAddressObservable, sendAmountObservable, accountPasswordObservable, sendObservable);
	}
}

File: app/src/main/kotlin/com/github/willjgriff/ethereumwallet/ui/screens/send/SendController.kt
package com.github.willjgriff.ethereumwallet.ui.screens.send

import android.support.v7.app.AlertDialog
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.github.willjgriff.ethereumwallet.R
import com.github.willjgriff.ethereumwallet.mvp.BaseMvpControllerKotlin
import com.github.willjgriff.ethereumwallet.ui.navigation.NavigationToolbarListener
import com.github.willjgriff.ethereumwallet.ui.screens.send.di.injectPresenter
import com.github.willjgriff.ethereumwallet.ui.screens.send.mvp.SendPresenter
import com.github.willjgriff.ethereumwallet.ui.screens.send.mvp.SendView
import com.jakewharton.rxbinding2.view.RxView
import kotlinx.android.synthetic.main.controller_send.view.*
import javax.inject.Inject

/**
 * Created by Will on 29/01/2017.
 */
class SendController : BaseMvpControllerKotlin<SendView, SendPresenter>(),
        SendView {

    override val mvpView: SendView get() = this
    @Inject lateinit override var presenter: SendPresenter

    init {
        injectPresenter()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup): View {
        val view = inflater.inflate(R.layout.controller_send, container, false)
        setupToolbarTitle()
        setupObservables(view)
        return view
    }

    private fun setupToolbarTitle() {
        if (targetController is NavigationToolbarListener) {
            (targetController as NavigationToolbarListener)
                    .setToolbarTitle(applicationContext!!.getString(R.string.controller_send_title))
        }
    }

    private fun setupObservables(view: View) {
        val recipientAddress = view.controller_send_recipient_address
        val sendAmount = view.controller_send_amount
        val accountPassword = view.controller_send_password
        val sendEtherClicked = RxView.clicks(view.controller_send_send_ether).share()

        recipientAddress.setCheckValidationTrigger(sendEtherClicked)
        sendAmount.setCheckValidationTrigger(sendEtherClicked)
        accountPassword.setCheckValidationTrigger(sendEtherClicked)

        presenter.apply {
            recipientChanged = recipientAddress.textChangedObservable
            recipientValid = recipientAddress.textValidObservable
            amountChanged = sendAmount.textChangedObservable
            amountValid = sendAmount.textValidObservable
            passwordChanged = accountPassword.textChangedObservable
            passwordValid = accountPassword.textValidObservable
            sendClicked = sendEtherClicked
        }
    }

    override fun displayTransactionSubmitted() {
        activity?.let {
            AlertDialog.Builder(it)
                    .setTitle(applicationContext?.getString(R.string.controller_send_success_dialog_title))
                    .setMessage(applicationContext?.getString(R.string.controller_send_success_dialog_message))
                    .setPositiveButton(applicationContext?.getString(R.string.common_ok), { dialog, _ -> dialog.dismiss() })
                    .show()
        }
    }

    override fun displayErrorSubmittingTx() {
        activity?.let {
            AlertDialog.Builder(it)
                    .setTitle(applicationContext?.getString(R.string.controller_send_error_dialog_title))
                    .setMessage(applicationContext?.getString(R.string.controller_send_error_dialog_message))
                    .setPositiveButton(applicationContext?.getString(R.string.common_ok), { dialog, _ -> dialog.dismiss() })
                    .show()
        }
    }

    override fun setBalance(balance: String) {
        view?.controller_send_current_balance?.text = balance
    }

    override fun setPendingBalance(pendingBalance: String) {
        view?.controller_send_pending_balance?.text = pendingBalance
    }
}
